bpo-40422: move _Py_closerange to core

kevans91 opened the pull request:
This API is relatively lightweight and organizationally, given that it's
used by multiple modules, it makes sense to move it to fileutils.


<!-- issue-number: [bpo-40422](https://bugs.python.org/issue40422) -->
https://bugs.python.org/issue40422
<!-- /issue-number -->

kevans91 committed:
bpo-40422: move _Py_closerange to core

This API is relatively lightweight and organizationally, given that it's
used by multiple modules, it makes sense to move it to fileutils.

vstinner commented on Include/fileutils.h:
```diff
@@ -18,6 +18,8 @@ PyAPI_FUNC(char*) _Py_EncodeLocaleRaw(
     size_t *error_pos);
 #endif
 
+PyAPI_FUNC(void) _Py_closerange(int first, int last);
```
Since it's a private function, I would prefer to define it in Include/internal/pycore_fileutils.h.

If you consider that it's a really important function, make it public (remove "_" prefix). I would prefer to declare it as an internal function.

kevans91 replied:
I had originally started going that route, but either subprocess of posixmodule (I want to say subprocess) doesn't get built in the needed context for that (from memory, that header required some kind of pycore define).

kevans91 committed:
Move _Py_closerange to the proper spot for internal API

Requires making sure that _posixsubprocess is compiled with the appropriate
Py_BUIILD_CORE_BUILTIN macro.

vstinner commented on Python/fileutils.c:
```diff
@@ -2106,3 +2106,79 @@ _Py_GetLocaleconvNumeric(struct lconv *lc,
+
+#ifdef HAVE_FDWALK
+#define USE_FDWALK
+#endif /* HAVE_FDWALK */
```
Nitipick: I like to indent for readability.

```suggestion
#ifdef __FreeBSD__
#  define USE_CLOSEFROM
#endif /* __FreeBSD__ */

#ifdef HAVE_FDWALK
#  define USE_FDWALK
#endif /* HAVE_FDWALK */
```

vstinner reviewed (approved):
LGTM, I just have a minor coding style suggestion.

kevans91 committed:
Incorporate indentation change

Co-authored-by: Victor Stinner <vstinner@python.org>
